File: languages/swift/binary-trees.swift
import Foundation

final class Node {
    let left: Node?
    let right: Node?
    init(_ left: Node?, _ right: Node?) {
        self.left = left
        self.right = right
    }
}

func make(_ depth: Int) -> Node {
    if depth == 0 {
        return Node(nil, nil)
    }
    return Node(make(depth - 1), make(depth - 1))
}

func check(_ node: Node) -> Int {
    if node.left == nil {
        return 1
    }
    return 1 + check(node.left!) + check(node.right!)
}

func run(_ n: Int) -> Int {
    let minDepth = 4
    let maxDepth = max(minDepth + 2, n)
    let stretchDepth = maxDepth + 1

    var total = check(make(stretchDepth))
    let longLived = make(maxDepth)

    var depth = minDepth
    while depth <= maxDepth {
        let iterations = 1 << (maxDepth - depth + minDepth)
        var s = 0
        for _ in 0..<iterations {
            s += check(make(depth))
        }
        total += s
        depth += 2
    }

    total += check(longLived)
    return total
}

let n = CommandLine.arguments.count > 1 ? (Int(CommandLine.arguments[1]) ?? 10) : 10
print(run(n))
print("binary-trees(\(n))")

File: languages/rust/binary-trees.rs
use std::env;

struct Node {
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn make(depth: i32) -> Box<Node> {
    if depth == 0 {
        Box::new(Node { left: None, right: None })
    } else {
        Box::new(Node {
            left: Some(make(depth - 1)),
            right: Some(make(depth - 1)),
        })
    }
}

fn check(node: &Node) -> i64 {
    match &node.left {
        None => 1,
        Some(l) => 1 + check(l) + check(node.right.as_ref().unwrap()),
    }
}

fn run(n: i32) -> i64 {
    let min_depth = 4;
    let max_depth = std::cmp::max(min_depth + 2, n);
    let stretch_depth = max_depth + 1;

    let mut total = check(&make(stretch_depth));
    let long_lived = make(max_depth);

    let mut depth = min_depth;
    while depth <= max_depth {
        let iterations = 1i64 << (max_depth - depth + min_depth);
        let mut s = 0i64;
        for _ in 0..iterations {
            s += check(&make(depth));
        }
        total += s;
        depth += 2;
    }

    total += check(&long_lived);
    total
}

fn main() {
    let n: i32 = env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(10);
    println!("{}", run(n));
    println!("binary-trees({})", n);
}
